test/jdk/java/awt/im/4959409/bug4959409.java

```diff
@@ -22,51 +22,180 @@
+        final CountDownLatch keyPressedEventLatch = new CountDownLatch(1);
+        final Point points[] = new Point[1];
+        final Rectangle rect[] = new Rectangle[1];
+
+        SwingUtilities.invokeAndWait(() -> {
+            frame = new JFrame("Test bug4959409");
+            jTextField = new JTextField();
+            jLabel = new JLabel();
+            frame.setLayout(new BorderLayout());
+            frame.addComponentListener(new ComponentAdapter() {
+                @Override
+                public void componentShown(ComponentEvent e) {
+                    super.componentShown(e);
+                    System.out.println("Frame is visible " + e.toString());
+                    frameVisibleLatch.countDown();
+                }
+            });
+
+            jTextField.addComponentListener(new ComponentAdapter() {
+                @Override
+                public void componentShown(ComponentEvent e) {
+                    super.componentShown(e);
+                    System.out.println("Component is visible + " + e.toString());
+                    componentVisibleLatch.countDown();
+                }
+            });
 
-public class bug4959409 extends javax.swing.JApplet {
-    public void init() {
-        new TestFrame();
+            jTextField.addKeyListener(new KeyAdapter() {
+
+                @Override
+                public void keyPressed(KeyEvent keyEvent) {
+                    super.keyPressed(keyEvent);
+                    int code = keyEvent.getKeyCode();
+                    int mod = keyEvent.getModifiersEx();
+                    if (code == '1' && mod == KeyEvent.SHIFT_DOWN_MASK) {
+                        keyPressedEventLatch.countDown();
+                        jLabel.setText("KEYPRESS received for Shift+1");
+                        System.out.println("KEYPRESS received for Shift+1");
+                    } else {
+                        jLabel.setText("Did not received KEY PRESS for Shift+1");
+                        System.out.println("Did not received KEY PRESS for Shift+1");
+                    }
+                }
+            });
+
+            Container container = frame.getContentPane();
+            container.add(jTextField, BorderLayout.SOUTH);
+            container.add(jLabel, BorderLayout.CENTER);
+            frame.setSize(300, 300);
+            frame.setLocationRelativeTo(null);
+            frame.setAlwaysOnTop(true);
+            frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+            frame.setVisible(true);
+        });
+
+        boolean isFrameVisible = frameVisibleLatch.await(TIMEOUT, TimeUnit.SECONDS);
+        if (!isFrameVisible) {
+            throw new RuntimeException("Frame is not visible after " + TIMEOUT + "  seconds");
+        }
+
+        if (!isComponentVisible(jTextField, componentVisibleLatch)) {
+            throw new RuntimeException("Component is not visible after " + TIMEOUT + "   seconds");
+        }
+
+        Robot robot = new Robot();
```

Comment:
Won't using `robot.setAutoDelay(DELAY)` give the same effect without the need to interleave key presses and mouse moves with explicit delay?

Reply:
added